In [ ]:
#0-1　基本設定

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as ticker
import seaborn as sns
import warnings
!pip install catboost

%matplotlib inline
%config Inlinebackend.figure_format='retina'
sns.set(style='whitegrid')
warnings.filterwarnings('ignore')
pd.set_option('display.max_columns',100)

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#0-2 データ読込
train=pd.read_parquet('/content/drive/MyDrive/Colab Notebooks/Kaggle/home-credit-default-risk/加工データ/01クリーニング/application_train_cleaning.zstd.parquet')
test =pd.read_parquet('/content/drive/MyDrive/Colab Notebooks/Kaggle/home-credit-default-risk/加工データ/01クリーニング/application_test_cleaning.zstd.parquet')

instalments_payments=pd.read_csv('/content/drive/MyDrive/Colab Notebooks/Kaggle/home-credit-default-risk/原データ/installments_payments.csv')

print(f"Memory Usage: {train.memory_usage().sum() / 1024**2:.2f} MB")
train.head()

Memory Usage: 846.67 MB


,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,NAME_TYPE_SUITE,NAME_INCOME_TYPE,NAME_EDUCATION_TYPE,NAME_FAMILY_STATUS,NAME_HOUSING_TYPE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,DAYS_REGISTRATION,DAYS_ID_PUBLISH,OWN_CAR_AGE,FLAG_MOBIL,FLAG_EMP_PHONE,FLAG_WORK_PHONE,FLAG_CONT_MOBILE,FLAG_PHONE,FLAG_EMAIL,OCCUPATION_TYPE,CNT_FAM_MEMBERS,REGION_RATING_CLIENT,REGION_RATING_CLIENT_W_CITY,WEEKDAY_APPR_PROCESS_START,HOUR_APPR_PROCESS_START,REG_REGION_NOT_LIVE_REGION,REG_REGION_NOT_WORK_REGION,LIVE_REGION_NOT_WORK_REGION,REG_CITY_NOT_LIVE_CITY,REG_CITY_NOT_WORK_CITY,LIVE_CITY_NOT_WORK_CITY,ORGANIZATION_TYPE,EXT_SOURCE_1,EXT_SOURCE_2,EXT_SOURCE_3,APARTMENTS_AVG,BASEMENTAREA_AVG,YEARS_BEGINEXPLUATATION_AVG,YEARS_BUILD_AVG,COMMONAREA_AVG,ELEVATORS_AVG,...,CC_CNT_DRAWINGS_OTHER_CURRENT_min,CC_CNT_DRAWINGS_OTHER_CURRENT_max,CC_CNT_DRAWINGS_OTHER_CURRENT_mean,CC_CNT_DRAWINGS_OTHER_CURRENT_sum,CC_CNT_DRAWINGS_OTHER_CURRENT_std,CC_CNT_DRAWINGS_POS_CURRENT_min,CC_CNT_DRAWINGS_POS_CURRENT_max,CC_CNT_DRAWINGS_POS_CURRENT_mean,CC_CNT_DRAWINGS_POS_CURRENT_sum,CC_CNT_DRAWINGS_POS_CURRENT_std,CC_CNT_INSTALMENT_MATURE_CUM_min,CC_CNT_INSTALMENT_MATURE_CUM_max,CC_CNT_INSTALMENT_MATURE_CUM_mean,CC_CNT_INSTALMENT_MATURE_CUM_sum,CC_CNT_INSTALMENT_MATURE_CUM_std,CC_SK_DPD_min,CC_SK_DPD_max,CC_SK_DPD_mean,CC_SK_DPD_sum,CC_SK_DPD_std,CC_SK_DPD_DEF_min,CC_SK_DPD_DEF_max,CC_SK_DPD_DEF_mean,CC_SK_DPD_DEF_sum,CC_SK_DPD_DEF_std,CC_NAME_CONTRACT_STATUS_nunique,CC_SK_ID_PREV_latest,CC_AMT_BALANCE_latest,CC_AMT_CREDIT_LIMIT_ACTUAL_latest,CC_AMT_DRAWINGS_ATM_CURRENT_latest,CC_AMT_DRAWINGS_CURRENT_latest,CC_AMT_DRAWINGS_OTHER_CURRENT_latest,CC_AMT_DRAWINGS_POS_CURRENT_latest,CC_AMT_INST_MIN_REGULARITY_latest,CC_AMT_PAYMENT_CURRENT_latest,CC_AMT_PAYMENT_TOTAL_CURRENT_latest,CC_AMT_RECEIVABLE_PRINCIPAL_latest,CC_AMT_RECIVABLE_latest,CC_AMT_TOTAL_RECEIVABLE_latest,CC_CNT_DRAWINGS_ATM_CURRENT_latest,CC_CNT_DRAWINGS_CURRENT_latest,CC_CNT_DRAWINGS_OTHER_CURRENT_latest,CC_CNT_DRAWINGS_POS_CURRENT_latest,CC_CNT_INSTALMENT_MATURE_CUM_latest,CC_SK_DPD_latest,CC_SK_DPD_DEF_latest,CC_NAME_CONTRACT_STATUS_latest,CC_MONTHS_BALANCE_min,CC_MONTHS_BALANCE_max,CC_MONTHS_BALANCE_count
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,351000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.018801,-9461,-637,-3648.0,-2120,NaN,1,1,0,1,1,0,Laborers,1.0,2,2,WEDNESDAY,10,0,0,0,0,0,0,Business Entity Type 3,0.083037,0.262949,0.139376,0.0247,0.0369,0.9722,0.6192,0.0143,0.00,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,1129500.0,Family,State servant,Higher education,Married,House / apartment,0.003541,-16765,-1188,-1186.0,-291,NaN,1,1,0,1,1,0,Core staff,2.0,1,1,MONDAY,11,0,0,0,0,0,0,School,0.311267,0.622246,NaN,0.0959,0.0529,0.9851,0.7960,0.0605,0.08,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,135000.0,Unaccompanied,Working,Secondary / secondary special,Single / not married,House / apartment,0.010032,-19046,-225,-4260.0,-2531,26.0,1,1,1,1,1,0,Laborers,1.0,2,2,MONDAY,9,0,0,0,0,0,0,Government,NaN,0.555912,0.729567,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,100006,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,297000.0,Unaccompanied,Working,Secondary / secondary special,Civil marriage,House / apartment,0.008019,-19005,-3039,-9833

In [ ]:
'''
FE方針まとめ
EDA方針メモ
1 基本EDA (均衡度、欠損確認、統計量)
　54カラムに対して以下を実施。
・異常値フラグ(Max/95％タイル値＞100)
　abnormal_MAX_frag(上記1.が100を超えるものを'1'、その他'0')
　※テストデータへの適用時に95％タイル値はトレインデータの値を使用すること。
2 ターゲット変数(返済Y回目、遅延X日)
　追加なし
3 試走結果に基づくEDA(第1有力指標)

・平均
　EXT_SOURCE_1～3の平均

・平均(分母補正あり)
　分子式　EXT_SOURCE_1～3の合計
　分母式　2＋(EXT_SOURCE_2(欠損の場合は0))
　補足：EXT_SOURCE_1、3については欠損時のデフォルト率が高いため、欠損時でも母数に含める

・スコアごとのデフォルト率
　スコアを0～1まで0.05刻みでデフォルト率を計算し、各スコアに紐づけ
　補足：グラフ上のデフォルトと非デフォルトのグラフの開差をモデルに認識させるため
　※リーク対策：Out-of-Fold＋Smoothing

・幾何平均
　EXT_SOURCE_1～3の乗算の3乗根
　※欠損時は各EXT_SOURCEごとにデフォルト率が近いスコアを適用
　補足：各EXT_SOURCE間にばらつきがある場合の特徴を捉えるため追加

・境界値フラグ(EXT_SOURCE_1～3)
　※境界値未満を'1'、境界値以上を'0'
　また、ターゲット値そのものを使うわけではないので、OOF＋Smoothingは実施しない。 --- EXT_SOURCE Border Lines ---
　EXT_SOURCE_1: 0.4284
　EXT_SOURCE_2: 0.4895
　EXT_SOURCE_3: 0.4484

・業種別のデフォルト率(ORGANIZATION_TYPE)
　※リーク対策：Out-of-Fold＋Smoothing
　※組み合わせ特徴量は以降のEDA時に有効なものがあれば追加。

4 試走結果に基づくEDA(第2有力指標)
・ビン(50)別デフォルト率
　※リーク対策：Out-of-Fold＋Smoothing
　AMT_ANNUITY
　INS_AMT_PAYMENT_min
　BU_DAYS_CREDIT_max
　AMT_CREDIT
　DAYS_BIRTH
　AMT_GOODS_PRICE

・キリ番(AMT_CREDIT,AMT_GOODS_PRICE)
　一定間隔で件数が増えることから、le4,le5,le6別で除算した数に小数点以下を含まないものについてフラグを追加

5 モデル試走(LGBM,CAT_K-FOLD)
-　評価方法をK-Foldに変更、baseline再計測
-　手順4までのカラムを追加し、再試走

6 パーケ出力
'''

"\nFE方針まとめ\nEDA方針メモ\n1 基本EDA (均衡度、欠損確認、統計量)\n\u300054カラムに対して以下を実施。\n・異常値フラグ(Max/95％タイル値＞100)\n\u3000abnormal_MAX_frag(上記1.が100を超えるものを'1'、その他'0')\n\u3000※テストデータへの適用時に95％タイル値はトレインデータの値を使用すること。\n2 ターゲット変数(返済Y回目、遅延X日)\n\u3000追加なし\n3 試走結果に基づくEDA(第1有力指標)\n\n・平均\n\u3000EXT_SOURCE_1～3の平均\n\n・平均(分母補正あり)\n\u3000分子式\u3000EXT_SOURCE_1～3の合計\n\u3000分母式\u30002＋(EXT_SOURCE_2(欠損の場合は0))\n\u3000補足：EXT_SOURCE_1、3については欠損時のデフォルト率が高いため、欠損時でも母数に含める\n\n・スコアごとのデフォルト率\n\u3000スコアを0～1まで0.05刻みでデフォルト率を計算し、各スコアに紐づけ\n\u3000補足：グラフ上のデフォルトと非デフォルトのグラフの開差をモデルに認識させるため\n\u3000※リーク対策：Out-of-Fold＋Smoothing\n\n・幾何平均\n\u3000EXT_SOURCE_1～3の乗算の3乗根\n\u3000※欠損時は各EXT_SOURCEごとにデフォルト率が近いスコアを適用\n\u3000補足：各EXT_SOURCE間にばらつきがある場合の特徴を捉えるため追加\n\n・境界値フラグ(EXT_SOURCE_1～3)\n\u3000※境界値未満を'1'、境界値以上を'0'\n\u3000また、ターゲット値そのものを使うわけではないので、OOF＋Smoothingは実施しない。 --- EXT_SOURCE Border Lines ---\n\u3000EXT_SOURCE_1: 0.4284\n\u3000EXT_SOURCE_2: 0.4895\n\u3000EXT_SOURCE_3: 0.4484\u3000\n\n・業種別のデフォルト率(ORGANIZATION_TYPE)\n\u3000※リーク対策：Out-of-Fold＋Smoothing\n\u3

In [ ]:
#1-1　異常値フラグ(Max/95％タイル値＞100)

outlier_cols = [
    'BU_AMT_CREDIT_MAX_OVERDUE_min', 'CC_SK_DPD_DEF_sum', 'POS_SK_DPD_DEF_sum',
    'CC_SK_DPD_DEF_mean', 'BU_AMT_CREDIT_MAX_OVERDUE_mean', 'POS_SK_DPD_sum',
    'CC_SK_DPD_sum', 'PREV_SELLERPLACE_AREA_min', 'POS_SK_DPD_DEF_mean',
    'BU_AMT_CREDIT_MAX_OVERDUE_std', 'BU_AMT_CREDIT_MAX_OVERDUE_max',
    'BU_AMT_CREDIT_MAX_OVERDUE_sum', 'CC_SK_DPD_mean', 'CC_SK_DPD_DEF_max',
    'CC_SK_DPD_DEF_std', 'PREV_SELLERPLACE_AREA_mean', 'PREV_SELLERPLACE_AREA_latest',
    'POS_SK_DPD_mean', 'PREV_DAYS_LAST_DUE_1ST_VERSION_min', 'BU_AMT_CREDIT_MAX_OVERDUE_latest',
    'PREV_SELLERPLACE_AREA_max', 'BU_AMT_ANNUITY_latest', 'BU_AMT_ANNUITY_std',
    'PREV_SELLERPLACE_AREA_std', 'BU_AMT_ANNUITY_sum', 'BU_AMT_ANNUITY_max',
    'BU_AMT_ANNUITY_min', 'POS_SK_DPD_DEF_std', 'PREV_SELLERPLACE_AREA_sum',
    'BU_AMT_ANNUITY_mean', 'POS_SK_DPD_std', 'POS_SK_DPD_DEF_max',
    'BU_AMT_CREDIT_SUM_min', 'BU_AMT_CREDIT_SUM_DEBT_min', 'CC_SK_DPD_max',
    'AMT_INCOME_TOTAL', 'CC_SK_DPD_std', 'CC_AMT_INST_MIN_REGULARITY_min',
    'POS_SK_DPD_max', 'BU_AMT_CREDIT_SUM_LIMIT_latest', 'BU_AMT_CREDIT_SUM_std',
    'BU_AMT_CREDIT_SUM_mean', 'CC_AMT_DRAWINGS_OTHER_CURRENT_mean', 'BU_AMT_CREDIT_SUM_sum',
    'INS_AMT_PAYMENT_min', 'CC_AMT_DRAWINGS_OTHER_CURRENT_sum', 'BU_BB_STATUS_NUM_MEAN_sum',
    'BU_AMT_CREDIT_SUM_LIMIT_mean', 'AMT_REQ_CREDIT_BUREAU_QRT', 'BU_AMT_CREDIT_SUM_DEBT_sum',
    'INS_AMT_INSTALMENT_min', 'PREV_AMT_DOWN_PAYMENT_min', 'BU_AMT_CREDIT_SUM_max',
    'CC_AMT_PAYMENT_CURRENT_min'
]

def apply_abnormal_flags(train_df, test_df, cols):
    # 1. 必要なカラムだけ抽出（存在しない場合は無視）
    target_cols = [c for c in cols if c in train_df.columns]

    # 2. フラグだけを格納する辞書
    train_new_flags = {}
    test_new_flags = {}

    for col in target_cols:
        # 計算時だけ float64 として扱う
        series_train = train_df[col].astype(float)
        q95_train = series_train.quantile(0.95)

        new_col_name = f'{col}_abnormal_MAX_frag'
        divisor = q95_train + 1e-9

        # 判定後、即座に int8 (メモリ最小) に変換
        train_new_flags[new_col_name] = (series_train / divisor > 100).astype('int8')
        test_new_flags[new_col_name] = (test_df[col].astype(float) / divisor > 100).astype('int8')

    # 3. 最後に一括結合（メモリ負荷を最小化）
    train_df = pd.concat([train_df, pd.DataFrame(train_new_flags, index=train_df.index)], axis=1)
    test_df = pd.concat([test_df, pd.DataFrame(test_new_flags, index=test_df.index)], axis=1)

    return train_df, test_df
print(len(outlier_cols))

54


In [ ]:
#2-1 ターゲット変数(返済Y回目、遅延X日)
#追加なし

In [ ]:
#3-1 FE(試走結果に基づくEDA(第1有力指標))　EXT_SOURCE_1～3の平均

def apply_ext_source_mean(df):
    ext_src_cols = ['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']
    # 3つのスコアの平均（欠損値を除いて計算）
    df['EXT_SOURCES_MEAN'] = df[ext_src_cols].mean(axis=1)

    return df


In [ ]:
#3-2 FE(試走結果に基づくEDA(第1有力指標))　EXT_SOURCE_1～3の平均(分母補正あり)

def apply_custom_ext_mean(df):

    # EXT_SOURCE_1～3の特殊補正平均
    # 分子：1〜3の合計（欠損は0扱い）
    # 分母：2 + (EXT_2が存在すれば1、なければ0)

    # 欠損を0に置換して合計を計算
    ext_sum = df[['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']].fillna(0).sum(axis=1)

    # 分母の計算: EXT_SOURCE_2が欠損していなければ1、していれば0を返す
    ext2_notnull = df['EXT_SOURCE_2'].notnull().astype(int)
    denominator = 2 + ext2_notnull

    # 特殊平均の算出
    df['EXT_SOURCES_CUSTOM_MEAN'] = ext_sum / denominator

    return df

In [ ]:
#3-3 FE(試走結果に基づくEDA(第1有力指標))　スコアごとのデフォルト率(EXT_SOURCE_1～3)

from sklearn.model_selection import KFold

def apply_ext_score_bin_te(train_df, test_df, target_col='TARGET', n_folds=5, smoothing=10):
    train_res = train_df.copy()
    test_res = test_df.copy()

    col = 'EXT_SOURCES_CUSTOM_MEAN'
    new_col = 'EXT_SOURCES_BIN_TARGET_RATE'

    bins = np.arange(0, 1.05, 0.05)

    train_res['ext_bin'] = pd.cut(train_res[col], bins=bins, include_lowest=True, labels=False)
    test_res['ext_bin'] = pd.cut(test_res[col], bins=bins, include_lowest=True, labels=False)

    kf = KFold(n_splits=n_folds, shuffle=True, random_state=42)
    global_mean = train_res[target_col].mean()

    oof_values = np.zeros(len(train_res), dtype=np.float64)

    for train_idx, val_idx in kf.split(train_res):
        df_t = train_res.iloc[train_idx]
        df_v = train_res.iloc[val_idx]

        agg = df_t.groupby('ext_bin')[target_col].agg(['count', 'mean'])
        counts = agg['count']
        means = agg['mean']
        smooth = (means * counts + global_mean * smoothing) / (counts + smoothing)

        oof_values[val_idx] = df_v['ext_bin'].map(smooth).values

    oof_values = np.where(np.isnan(oof_values), global_mean, oof_values)
    train_res[new_col] = oof_values

    agg = train_res.groupby('ext_bin')[target_col].agg(['count', 'mean'])
    smooth = (agg['mean'] * agg['count'] + global_mean * smoothing) / (agg['count'] + smoothing)
    test_values = test_res['ext_bin'].map(smooth).values.astype(np.float64)
    test_values = np.where(np.isnan(test_values), global_mean, test_values)
    test_res[new_col] = test_values

    train_res = train_res.drop(columns=['ext_bin'])
    test_res = test_res.drop(columns=['ext_bin'])

    return train_res, test_res

In [ ]:
#3-4(1)　FE(試走結果に基づくEDA(第1有力指標))　幾何平均(EXT_SOURCE_1～3)
def analyze_ext_nan_substitute(df):
    ext_cols = ['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']

    for col in ext_cols:
        print(f"--- {col} の調査 ---")

        # 1. 有効値があるデータのデフォルト率を0.05刻みで計算
        temp = df[[col, 'TARGET']].dropna(subset=[col]).copy()
        # ビン化（0.05刻み）
        bins = np.arange(0, 1.05, 0.05)
        temp['bin'] = pd.cut(temp[col], bins=bins, include_lowest=True)

        # スコア帯ごとのデフォルト率
        bin_stats = temp.groupby('bin')['TARGET'].mean()

        # 2. NaNデータのデフォルト率
        nan_default_rate = df[df[col].isnull()]['TARGET'].mean()

        # 3. 最もデフォルト率が近いスコア帯を特定
        # 差分の絶対値が最小のインデックスを取得
        closest_bin = (bin_stats - nan_default_rate).abs().idxmin()
        closest_bin_rate = bin_stats[closest_bin]

        # 指定の形式で表示
        print(f"NaNデータのデフォルト率: {nan_default_rate:.4f}")
        print(f"最もデフォルト率が近いスコア帯: {closest_bin}")
        print(f"最もデフォルト率が近いスコア帯のデフォルト率: {closest_bin_rate:.4f}")
        print(f"※推奨される補完用スコア値 (中央値): {closest_bin.mid:.4f}")
        print("\n")

# 実行（trainデータを使用）
analyze_ext_nan_substitute(train)

--- EXT_SOURCE_1 の調査 ---
NaNデータのデフォルト率: 0.0852
最もデフォルト率が近いスコア帯: (0.35, 0.4]
最もデフォルト率が近いスコア帯のデフォルト率: 0.0891
※推奨される補完用スコア値 (中央値): 0.3750


--- EXT_SOURCE_2 の調査 ---
NaNデータのデフォルト率: 0.0788
最もデフォルト率が近いスコア帯: (0.5, 0.55]
最もデフォルト率が近いスコア帯のデフォルト率: 0.0776
※推奨される補完用スコア値 (中央値): 0.5250


--- EXT_SOURCE_3 の調査 ---
NaNデータのデフォルト率: 0.0931
最もデフォルト率が近いスコア帯: (0.35, 0.4]
最もデフォルト率が近いスコア帯のデフォルト率: 0.0936
※推奨される補完用スコア値 (中央値): 0.3750




上記より、補完値は以下の値で実施。  

*   EXT_SOURCE_1:0.3750
*   EXT_SOURCE_2:0.5250
*   EXT_SOURCE_3:0.3750

In [ ]:
#3-4(2)　FE(試走結果に基づくEDA(第1有力指標))　幾何平均(EXT_SOURCE_1～3)

def apply_ext_source_geometric_mean(train_df, test_df):
    """
    EXT_SOURCE_1～3の幾何平均（3乗根）を生成
    欠損時は、NaNデータのデフォルト率に基づいた特定スコアで補完
    """
    train_res = train_df.copy()
    test_res = test_df.copy()

    # 調査結果に基づく補完値の設定
    fill_values = {
        'EXT_SOURCE_1': 0.3750,
        'EXT_SOURCE_2': 0.5250,
        'EXT_SOURCE_3': 0.3750
    }

    ext_cols = ['EXT_SOURCE_1', 'EXT_SOURCE_2', 'EXT_SOURCE_3']

    for col in ext_cols:
        fv = fill_values[col]
        # 欠損値を指定のスコアで補完した一時的なカラムを作成
        train_res[f'{col}_filled'] = train_res[col].fillna(fv)
        test_res[f'{col}_filled'] = test_res[col].fillna(fv)

    # 幾何平均の計算（3つの乗算の1/3乗）
    train_res['EXT_SOURCES_GEOM_MEAN'] = (
        train_res['EXT_SOURCE_1_filled'] * train_res['EXT_SOURCE_2_filled'] * train_res['EXT_SOURCE_3_filled']
    ) ** (1/3)

    test_res['EXT_SOURCES_GEOM_MEAN'] = (
        test_res['EXT_SOURCE_1_filled'] * test_res['EXT_SOURCE_2_filled'] * test_res['EXT_SOURCE_3_filled']
    ) ** (1/3)

    # 作業用カラムの削除
    drop_cols = [f'{col}_filled' for col in ext_cols]
    train_res = train_res.drop(columns=drop_cols)
    test_res = test_res.drop(columns=drop_cols)

    return train_res, test_res

In [ ]:
#3-5　FE(試走結果に基づくEDA(第1有力指標))　境界値フラグ(EXT_SOURCE_1～3)

def apply_ext_source_border_flags(df):

    # EXT_SOURCE_1～3の境界値フラグを作成
    # 境界値未満を '1'、境界値以上を '0' とする

    # 指定された境界値の設定
    borders = {
        'EXT_SOURCE_1': 0.4284,
        'EXT_SOURCE_2': 0.4895,
        'EXT_SOURCE_3': 0.4484
    }

    for col, threshold in borders.items():
        if col in df.columns:
            # 境界値未満なら1、以上なら0（欠損値は境界値以上ではないため0になります）
            df[f'{col}_border_frag'] = (df[col] < threshold).astype(int)

    return df

In [ ]:
#3-6　FE(試走結果に基づくEDA(第1有力指標))　業種別のデフォルト率(ORGANIZATION_TYPE)
from sklearn.model_selection import KFold

def apply_org_type_te(train_df, test_df, target_col='TARGET', n_folds=5, smoothing=10):
    train_res = train_df.copy()
    test_res = test_df.copy()

    col = 'ORGANIZATION_TYPE'
    new_col = 'ORG_TYPE_TARGET_RATE'

    kf = KFold(n_splits=n_folds, shuffle=True, random_state=42)
    global_mean = train_res[target_col].mean()

    # numpy配列に溜める
    oof_values = np.zeros(len(train_res), dtype=np.float64)

    for train_idx, val_idx in kf.split(train_res):
        df_t = train_res.iloc[train_idx]
        df_v = train_res.iloc[val_idx]

        agg = df_t.groupby(col)[target_col].agg(['count', 'mean'])
        counts = agg['count']
        means = agg['mean']
        smooth = (means * counts + global_mean * smoothing) / (counts + smoothing)

        # numpy配列に直接書き込み
        oof_values[val_idx] = df_v[col].map(smooth).values

    # 最後に一括代入
    train_res[new_col] = oof_values

    # Testデータ
    agg = train_res.groupby(col)[target_col].agg(['count', 'mean'])
    counts = agg['count']
    means = agg['mean']
    smooth = (means * counts + global_mean * smoothing) / (counts + smoothing)
    test_res[new_col] = test_res[col].map(smooth).values

    # oof_valuesの処理（forループの後）
    oof_values = np.where(np.isnan(oof_values), global_mean, oof_values)
    train_res[new_col] = oof_values

    # Testデータ
    agg = train_res.groupby(col)[target_col].agg(['count', 'mean'])
    counts = agg['count']
    means = agg['mean']
    smooth = (means * counts + global_mean * smoothing) / (counts + smoothing)
    test_values = test_res[col].map(smooth).values.astype(np.float64)
    test_values = np.where(np.isnan(test_values), global_mean, test_values)
    test_res[new_col] = test_values

    return train_res, test_res

In [ ]:
#4-1　FE(試走結果に基づくEDA(第2有力指標))　ビン(50)別デフォルト率

# 第2有力指標のリスト
secondary_importance_cols = [
    'AMT_ANNUITY',
    'INS_AMT_PAYMENT_min',
    'BU_DAYS_CREDIT_max',
    'AMT_CREDIT',
    'DAYS_BIRTH',
    'AMT_GOODS_PRICE'
]

from sklearn.model_selection import KFold

def apply_bin_target_encoding(train_df, test_df, target_cols, target_val='TARGET', n_bins=50, n_folds=5, smoothing=10):
    train_res = train_df.copy()
    test_res = test_df.copy()

    kf = KFold(n_splits=n_folds, shuffle=True, random_state=42)
    global_mean = train_res[target_val].mean()

    for col in target_cols:
        if col not in train_res.columns:
            continue

        new_col = f'{col}_BIN_TARGET_RATE'

        combined_min = train_res[col].min()
        combined_max = train_res[col].max()
        bins = np.linspace(combined_min, combined_max, n_bins + 1)

        train_res['tmp_bin'] = pd.cut(train_res[col], bins=bins, include_lowest=True, labels=False)
        test_res['tmp_bin'] = pd.cut(test_res[col], bins=bins, include_lowest=True, labels=False)

        oof_values = np.zeros(len(train_res), dtype=np.float64)

        for train_idx, val_idx in kf.split(train_res):
            df_t = train_res.iloc[train_idx]
            df_v = train_res.iloc[val_idx]

            agg = df_t.groupby('tmp_bin')[target_val].agg(['count', 'mean'])
            counts = agg['count']
            means = agg['mean']
            smooth = (means * counts + global_mean * smoothing) / (counts + smoothing)

            oof_values[val_idx] = df_v['tmp_bin'].map(smooth).values

        oof_values = np.where(np.isnan(oof_values), global_mean, oof_values)
        train_res[new_col] = oof_values

        agg = train_res.groupby('tmp_bin')[target_val].agg(['count', 'mean'])
        smooth = (agg['mean'] * agg['count'] + global_mean * smoothing) / (agg['count'] + smoothing)
        test_values = test_res['tmp_bin'].map(smooth).values.astype(np.float64)
        test_values = np.where(np.isnan(test_values), global_mean, test_values)
        test_res[new_col] = test_values

        train_res = train_res.drop(columns=['tmp_bin'])
        test_res = test_res.drop(columns=['tmp_bin'])

    return train_res, test_res


In [ ]:
#4-2　FE(試走結果に基づくEDA(第2有力指標))　キリ番(AMT_CREDIT,AMT_GOODS_PRICE)
def apply_round_number_flags(df):
    """
    AMT_CREDIT, AMT_GOODS_PRICE がキリの良い数字かどうかを判定
    1万, 10万, 100万で割り切れる（小数点以下がない）場合にフラグを立てる
    """
    cols = ['AMT_CREDIT', 'AMT_GOODS_PRICE']
    divisors = [10000, 100000, 1000000] # 1e4, 1e5, 1e6

    for col in cols:
        if col not in df.columns:
            continue

        for div in divisors:
            # 除算して余りが0（または非常に小さい誤差範囲）ならフラグ
            # divの大きさに合わせて 1e4なら "10k", 1e5なら "100k" のように命名
            label = f"{div/1000:g}k"
            new_col = f"{col}_is_round_{label}"

            # 欠損値は 0 (False) と判定されるように処理
            df[new_col] = (df[col] % div == 0).astype(int)

    return df

In [ ]:
#5-1　モデル試走(LGBM,CAT_K-FOLD)

import lightgbm as lgb
from catboost import CatBoostClassifier, Pool
from sklearn.model_selection import KFold
from sklearn.metrics import roc_auc_score

def train_lgbm(X, y, X_test, params=None):
    """LightGBMによるK-Fold学習"""
    if params is None:
        params = {
            'objective': 'binary', 'metric': 'auc', 'verbosity': -1,
            'learning_rate': 0.05, 'num_leaves': 31, 'random_state': 42
        }

    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    oof_preds = np.zeros(len(X))
    test_preds = np.zeros(len(X_test))

    for fold, (train_idx, val_idx) in enumerate(kf.split(X, y)):
        X_train, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_train, y_val = y.iloc[train_idx], y.iloc[val_idx]

        dtrain = lgb.Dataset(X_train, label=y_train)
        dval = lgb.Dataset(X_val, label=y_val, reference=dtrain)

        model = lgb.train(
            params, dtrain, valid_sets=[dtrain, dval],
            num_boost_round=10000,
            callbacks=[lgb.early_stopping(100), lgb.log_evaluation(0)]
        )

        oof_preds[val_idx] = model.predict(X_val)
        test_preds += model.predict(X_test) / kf.n_splits
        print(f"LGBM Fold {fold+1} AUC: {roc_auc_score(y_val, oof_preds[val_idx]):.5f}")

    print(f"LGBM Overall CV: {roc_auc_score(y, oof_preds):.5f}")
    return oof_preds, test_preds

def train_lgbm(X, y, X_test, params=None):
    if params is None:
        params = {'objective': 'binary', 'metric': 'auc', 'verbosity': -1, 'learning_rate': 0.05, 'num_leaves': 31, 'random_state': 42}
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    oof_preds = np.zeros(len(X))
    test_preds = np.zeros(len(X_test))
    models = [] # モデルを格納するリストを追加

    for fold, (train_idx, val_idx) in enumerate(kf.split(X, y)):
        X_train, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_train, y_val = y.iloc[train_idx], y.iloc[val_idx]
        dtrain = lgb.Dataset(X_train, label=y_train)
        dval = lgb.Dataset(X_val, label=y_val, reference=dtrain)

        model = lgb.train(params, dtrain, valid_sets=[dtrain, dval], num_boost_round=10000,
                          callbacks=[lgb.early_stopping(100), lgb.log_evaluation(0)])

        models.append(model) # リストに保存
        oof_preds[val_idx] = model.predict(X_val)
        test_preds += model.predict(X_test) / kf.n_splits
        print(f"LGBM Fold {fold+1} AUC: {roc_auc_score(y_val, oof_preds[val_idx]):.5f}")
    print(f"LGBM Overall CV: {roc_auc_score(y, oof_preds):.5f}")
    return oof_preds, test_preds, models # modelsを返す

def train_catboost(X, y, X_test, params=None):
    if params is None:
        params = {'loss_function': 'Logloss', 'eval_metric': 'AUC', 'learning_rate': 0.05, 'iterations': 10000, 'random_seed': 42, 'verbose': 0, 'early_stopping_rounds': 100}
    kf = KFold(n_splits=5, shuffle=True, random_state=42)
    oof_preds = np.zeros(len(X))
    test_preds = np.zeros(len(X_test))
    models = [] # モデルを格納するリストを追加

    for fold, (train_idx, val_idx) in enumerate(kf.split(X, y)):
        X_train, X_val = X.iloc[train_idx], X.iloc[val_idx]
        y_train, y_val = y.iloc[train_idx], y.iloc[val_idx]

        model = CatBoostClassifier(**params)
        model.fit(X_train, y_train, eval_set=(X_val, y_val), use_best_model=True)

        models.append(model) # リストに保存
        oof_preds[val_idx] = model.predict_proba(X_val)[:, 1]
        test_preds += model.predict_proba(X_test)[:, 1] / kf.n_splits
        print(f"CatBoost Fold {fold+1} AUC: {roc_auc_score(y_val, oof_preds[val_idx]):.5f}")
    print(f"CatBoost Overall CV: {roc_auc_score(y, oof_preds):.5f}")
    return oof_preds, test_preds, models # modelsを返す

In [ ]:
#5-2　モデル試走(LGBM,CAT_K-FOLD)
X = train.drop(columns=['TARGET', 'SK_ID_CURR']).select_dtypes(include='number')
y = train['TARGET']
X_test = test.drop(columns=['SK_ID_CURR']).select_dtypes(include='number')
X_test = X_test[X.columns]

# --- . モデルの実行 ---
oof_lgb_base, pred_lgb_base, models_lgb_base = train_lgbm(X, y, X_test)
oof_cat_base, pred_cat_base, models_cat_base = train_catboost(X, y, X_test)

KeyboardInterrupt: 

Baseline  
LGBM Overall CV: 0.77854  
CatBoost Overall CV: 0.78211  

In [ ]:
#5-3　アップキャスト(カラム追加前の加工用)
import gc

# 1-1以外のFE（3-1〜3-6等）で、計算元として参照されるカラム
fe_source_cols = [
    'TARGET',               # ターゲットエンコーディングの計算（最重要）
    'EXT_SOURCE_1',         # 3-5で使用
    'EXT_SOURCE_2',         # 3-5で使用
    'EXT_SOURCE_3',         # 3-5で使用
    'AMT_INCOME_TOTAL',     # 各種比率計算
    'AMT_CREDIT',
    'AMT_ANNUITY',
    'AMT_GOODS_PRICE',
    'DAYS_BIRTH'            # 3-1等で使用
]

# 現在のデータフレームに存在する数値型カラムのみを対象にする
target_cols = [c for c in fe_source_cols if c in train.columns]

# 全て float64 に変換して「器」を広げる
# (int型であっても計算結果が小数になるため、float64にするのが最も安全です)
train = train.astype({col: 'float64' for col in target_cols})
test = test.astype({col: 'float64' for col in target_cols if col in test.columns})

print(f"ターゲットエンコーディング等の計算用カラム {len(target_cols)} 個を float64 に変換")
gc.collect()

ターゲットエンコーディング等の計算用カラム 9 個を float64 に変換


0

In [ ]:
#5-4　カラム追加実行

#1-1　異常値フラグ(Max/95％タイル値＞100)
train, test = apply_abnormal_flags(train, test, outlier_cols)

#2-1　ターゲット変数
#追加なし

#3-1　FE(試走結果に基づくEDA(第1有力指標))　EXT_SOURCE_1～3の平均
train = apply_ext_source_mean(train)
test = apply_ext_source_mean(test)
#3-2　FE(試走結果に基づくEDA(第1有力指標))　EXT_SOURCE_1～3の平均(分母補正あり)
train = apply_custom_ext_mean(train)
test = apply_custom_ext_mean(test)
#3-3　FE(試走結果に基づくEDA(第1有力指標))　スコアごとのデフォルト率(EXT_SOURCE_1～3)
train, test = apply_ext_score_bin_te(train, test)
#3-4　FE(試走結果に基づくEDA(第1有力指標))　幾何平均(EXT_SOURCE_1～3)
train, test = apply_ext_source_geometric_mean(train, test)
#3-5　FE(試走結果に基づくEDA(第1有力指標))　境界値フラグ(EXT_SOURCE_1～3)
train = apply_ext_source_border_flags(train)
test = apply_ext_source_border_flags(test)
#3-6　FE(試走結果に基づくEDA(第1有力指標))　業種別のデフォルト率(ORGANIZATION_TYPE)
train, test = apply_org_type_te(train, test)

#4-1　FE(試走結果に基づくEDA(第2有力指標))　ビン(50)別デフォルト率
train, test = apply_bin_target_encoding(train, test, secondary_importance_cols)
#4-2　FE(試走結果に基づくEDA(第2有力指標))　キリ番(AMT_CREDIT,AMT_GOODS_PRICE)
train = apply_round_number_flags(train)
test = apply_round_number_flags(test)
gc.collect()

0

In [ ]:
#5-5　モデル試走(LGBM,CAT_K-FOLD)
X = train.drop(columns=['TARGET', 'SK_ID_CURR']).select_dtypes(include='number')
y = train['TARGET']
X_test = test.drop(columns=['SK_ID_CURR']).select_dtypes(include='number')
X_test = X_test[X.columns]

# --- . モデルの実行 ---
oof_lgb_fe, pred_lgb_fe, models_lgb_fe = train_lgbm(X, y, X_test)
oof_cat_fe, pred_cat_fe, models_cat_fe = train_catboost(X, y, X_test)

Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[289]	training's auc: 0.85701	valid_1's auc: 0.778357
LGBM Fold 1 AUC: 0.77836
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[656]	training's auc: 0.916447	valid_1's auc: 0.779893
LGBM Fold 2 AUC: 0.77989
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[670]	training's auc: 0.917109	valid_1's auc: 0.77788
LGBM Fold 3 AUC: 0.77788
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[372]	training's auc: 0.873029	valid_1's auc: 0.780169
LGBM Fold 4 AUC: 0.78017
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[467]	training's auc: 0.891965	valid_1's auc: 0.778099
LGBM Fold 5 AUC: 0.77810
LGBM Overall CV: 0.77875
CatBoost Fold 1 AUC: 0.78432
CatBoost Fold 2 AUC: 0.78466
CatBoost Fold 3 AUC: 0.78243
CatBo

FE1追加後  
LGBM Overall CV: 0.77875  
CatBoost Overall CV: 0.78359  

In [ ]:
#5-6　モデル寄与度

# --- 1. 寄与度の集計と正規化 (%) ---
def get_pct_importance(models, model_name, feature_names):
    if model_name == 'LGBM':
        # Gain（合計）を抽出して平均
        imp = np.mean([m.feature_importance(importance_type='gain') for m in models], axis=0)
    else:
        # CatBoostのImportanceを平均
        imp = np.mean([m.get_feature_importance() for m in models], axis=0)

    # 合計を100%に正規化
    imp_pct = (imp / np.sum(imp)) * 100
    return pd.DataFrame({'feature': feature_names, f'{model_name}_pct': imp_pct})

# 寄与度データ作成
lgb_imp_df = get_pct_importance(models_lgb_fe, 'LGBM', X.columns)
cat_imp_df = get_pct_importance(models_cat_fe, 'CAT', X.columns)

# --- 2. マージして比較表を作成 ---
comparison_df = pd.merge(lgb_imp_df, cat_imp_df, on='feature')

# AUCスコアの取得
lgb_auc = roc_auc_score(y, oof_lgb_fe)
cat_auc = roc_auc_score(y, oof_cat_fe)

# --- 3. 結果表示 ---
print(f" 精度比較 (AUC) :")
print(f"  - LightGBM: {lgb_auc:.4f}")
print(f"  - CatBoost: {cat_auc:.4f}\n")

# LGBMのTOP15
lgb_top15_s = comparison_df.sort_values(by='LGBM_pct', ascending=False).head(15)[['feature', 'LGBM_pct']]
lgb_top15_s.columns = ['LGBM Feature', 'Importance(%)']

# CatBoostのTOP15
cb_top15_s = comparison_df.sort_values(by='CAT_pct', ascending=False).head(15)[['feature', 'CAT_pct']]
cb_top15_s.columns = ['CatBoost Feature', 'Importance(%)']

# インデックスをリセットして横に並べる
lgb_top15_s = lgb_top15_s.reset_index(drop=True)
cb_top15_s = cb_top15_s.reset_index(drop=True)
side_by_side_s = pd.concat([lgb_top15_s, cb_top15_s], axis=1)

print(f" モデル別インポータンス TOP 15 ({len(X.columns)}列版)")
print(side_by_side_s.to_markdown(index=False))

# --- 共通 ---
common_top_s = set(lgb_top15_s['LGBM Feature']) & set(cb_top15_s['CatBoost Feature'])
print(f"\n 両モデルが共通してTOP15に挙げた項目: {len(common_top_s)}件")
print(sorted(list(common_top_s)))

 精度比較 (AUC) :
  - LightGBM: 0.7787
  - CatBoost: 0.7836

 モデル別インポータンス TOP 15 (653列版)
| LGBM Feature               |   Importance(%) | CatBoost Feature              |   Importance(%) |
|:---------------------------|----------------:|:------------------------------|----------------:|
| EXT_SOURCES_MEAN           |       16.5437   | EXT_SOURCES_GEOM_MEAN         |        9.2213   |
| EXT_SOURCES_GEOM_MEAN      |       12.9556   | EXT_SOURCES_MEAN              |        8.15575  |
| ORG_TYPE_TARGET_RATE       |        1.80701  | AMT_ANNUITY                   |        1.93987  |
| AMT_ANNUITY                |        1.33765  | EXT_SOURCES_CUSTOM_MEAN       |        1.92001  |
| INS_AMT_PAYMENT_min        |        1.27763  | ORG_TYPE_TARGET_RATE          |        1.8151   |
| EXT_SOURCE_3               |        1.21436  | DAYS_BIRTH                    |        1.58946  |
| DAYS_BIRTH                 |        1.07989  | EXT_SOURCE_3                  |        1.38394  |
| BU_DAYS_CREDIT_max    

FE1追加後  
LightGBM(ベースライン比)+0.00021  
CatBoost(ベースライン比)+0.00148  

寄与度の上位の様相は大きく変化したものの、AUC自体の精度は微増に留まった、これは寄与度の低い特徴量がノイズとなったことと、似た内容の特徴量であまり差がつけられなかったことが原因と思われる。

In [ ]:
#5-7　追加カラムの寄与度

# --- 追加カラムに特化したランキング表示コード ---

def show_added_features_ranking(models_lgb, models_cat, feature_names):
    # 1. LGBMの寄与度(%)とランク
    lgb_imp = np.mean([m.feature_importance(importance_type='gain') for m in models_lgb], axis=0)
    lgb_pct = (lgb_imp / np.sum(lgb_imp)) * 100

    # 2. CatBoostの寄与度(%)とランク
    cat_imp = np.mean([m.get_feature_importance() for m in models_cat], axis=0)
    cat_pct = (cat_imp / np.sum(cat_imp)) * 100

    # データフレーム化
    df = pd.DataFrame({
        'feature': feature_names,
        'LGBM_pct': lgb_pct,
        'CAT_pct': cat_pct
    })

    # 全体での順位を計算
    df['LGBM_Rank'] = df['LGBM_pct'].rank(ascending=False).astype(int)
    df['CAT_Rank'] = df['CAT_pct'].rank(ascending=False).astype(int)

    # 追加カラムのキーワードでフィルタリング
    added_keys = ['abnormal_MAX_frag', '_TE', '_mean', 'border', 'round', 'geometric']
    added_df = df[df['feature'].str.contains('|'.join(added_keywords))].copy()

    # 表示用に整形（CATの順位が高い順に並べる）
    added_df = added_df.sort_values(by='CAT_Rank')

    print(f"=== 追加カラム限定：寄与度(%) & 全体順位 (全{len(feature_names)}列中) ===")

    # カラム名をわかりやすくリネーム
    result = added_df[[
        'feature', 'LGBM_pct', 'LGBM_Rank', 'CAT_pct', 'CAT_Rank'
    ]].reset_index(drop=True)

    # 表形式で表示
    print(result.to_markdown(index=False))

    # 簡易サマリー
    top100_lgb = len(result[result['LGBM_Rank'] <= 100])
    top100_cat = len(result[result['CAT_Rank'] <= 100])
    print(f"\n[Summary]")
    print(f"TOP 100入りした追加カラム: LGBM {top100_lgb}個 / CatBoost {top100_cat}個")

# 実行
show_added_features_ranking(models_lgb_fe, models_cat_fe, X.columns)

=== 追加カラム限定：寄与度(%) & 全体順位 (全653列中) ===
| feature                                              |    LGBM_pct |   LGBM_Rank |     CAT_pct |   CAT_Rank |
|:-----------------------------------------------------|------------:|------------:|------------:|-----------:|
| EXT_SOURCE_3_border_frag                             | 0.000678499 |         562 | 0.777319    |         21 |
| BU_AMT_CREDIT_SUM_DEBT_mean                          | 0.54376     |          27 | 0.653662    |         24 |
| POS_CNT_INSTALMENT_FUTURE_mean                       | 0.597301    |          20 | 0.569271    |         27 |
| PREV_CNT_PAYMENT_mean                                | 0.63361     |          18 | 0.513769    |         31 |
| INS_AMT_PAYMENT_mean                                 | 0.463972    |          30 | 0.445982    |         41 |
| BU_AMT_CREDIT_SUM_mean                               | 0.407824    |          35 | 0.428764    |         42 |
| POS_SK_DPD_DEF_mean                                  | 0.560955

両モデルで寄与度0の追加したカラムを削除


In [ ]:
#5-8　不要カラム削除

# 1. 削除対象のカラムリスト（寄与度0だったもの）
cols_to_remove = [
    'PREV_SELLERPLACE_AREA_latest_abnormal_MAX_frag', 'PREV_SELLERPLACE_AREA_mean_abnormal_MAX_frag',
    'CC_SK_DPD_DEF_std_abnormal_MAX_frag', 'CC_SK_DPD_mean_abnormal_MAX_frag',
    'BU_AMT_CREDIT_MAX_OVERDUE_max_abnormal_MAX_frag', 'BU_AMT_CREDIT_MAX_OVERDUE_sum_abnormal_MAX_frag',
    'POS_SK_DPD_DEF_mean_abnormal_MAX_frag', 'BU_AMT_ANNUITY_std_abnormal_MAX_frag',
    'BU_AMT_ANNUITY_max_abnormal_MAX_frag', 'BU_AMT_ANNUITY_sum_abnormal_MAX_frag',
    'PREV_SELLERPLACE_AREA_std_abnormal_MAX_frag', 'POS_SK_DPD_DEF_std_abnormal_MAX_frag',
    'BU_AMT_ANNUITY_mean_abnormal_MAX_frag', 'PREV_SELLERPLACE_AREA_sum_abnormal_MAX_frag',
    'PREV_DAYS_LAST_DUE_1ST_VERSION_min_abnormal_MAX_frag', 'PREV_SELLERPLACE_AREA_max_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_min_abnormal_MAX_frag', 'POS_SK_DPD_DEF_max_abnormal_MAX_frag',
    'CC_SK_DPD_max_abnormal_MAX_frag', 'CC_SK_DPD_std_abnormal_MAX_frag',
    'POS_SK_DPD_max_abnormal_MAX_frag', 'BU_AMT_CREDIT_SUM_LIMIT_latest_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_std_abnormal_MAX_frag', 'BU_AMT_CREDIT_SUM_mean_abnormal_MAX_frag',
    'CC_AMT_DRAWINGS_OTHER_CURRENT_mean_abnormal_MAX_frag', 'BU_AMT_CREDIT_MAX_OVERDUE_std_abnormal_MAX_frag',
    'CC_SK_DPD_DEF_sum_abnormal_MAX_frag', 'BU_AMT_CREDIT_MAX_OVERDUE_min_abnormal_MAX_frag',
    'POS_SK_DPD_sum_abnormal_MAX_frag', 'CC_SK_DPD_sum_abnormal_MAX_frag',
    'CC_SK_DPD_DEF_mean_abnormal_MAX_frag', 'POS_SK_DPD_DEF_sum_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_DEBT_sum_abnormal_MAX_frag', 'AMT_REQ_CREDIT_BUREAU_QRT_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_LIMIT_mean_abnormal_MAX_frag', 'BU_BB_STATUS_NUM_MEAN_sum_abnormal_MAX_frag',
    'CC_AMT_DRAWINGS_OTHER_CURRENT_sum_abnormal_MAX_frag', 'INS_AMT_PAYMENT_min_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_sum_abnormal_MAX_frag', 'CC_AMT_PAYMENT_CURRENT_min_abnormal_MAX_frag',
    'INS_AMT_INSTALMENT_min_abnormal_MAX_frag', 'PREV_AMT_DOWN_PAYMENT_min_abnormal_MAX_frag',
    'BU_AMT_CREDIT_SUM_max_abnormal_MAX_frag', 'AMT_CREDIT_is_round_1000k', 'AMT_GOODS_PRICE_is_round_1000k'
]

# 2. 実際に存在するカラムだけを抽出して削除（エラー回避）
existing_drop_cols = [c for c in cols_to_remove if c in train.columns]
train = train.drop(columns=existing_drop_cols)
test = test.drop(columns=[c for c in existing_drop_cols if c in test.columns])

print(f"無効な追加カラム {len(existing_drop_cols)} 個を削除")

# 3. メモリのクリーンアップ
import gc
gc.collect()

無効な追加カラム 45 個を削除しました。


1074

In [ ]:
#5-9　モデル試走(LGBMのみ)

# 1. ターゲットと特徴量の再定義（不要カラム削除済みのtrain/testを使用）
X = train.drop(columns=['TARGET', 'SK_ID_CURR']).select_dtypes(include='number')
y = train['TARGET']
X_test = test.drop(columns=['SK_ID_CURR']).select_dtypes(include='number')
X_test = X_test[X.columns] # 列の順序を揃える

print(f"学習に使用するカラム数: {len(X.columns)}")

# 2. LGBMの実行（以前定義した関数を使用）
# 戻り値は models_lgb_slim として保存し、後で寄与度を見れるようにします
oof_lgb_slim, pred_lgb_slim, models_lgb_slim = train_lgbm(X, y, X_test)

# 3. スコアの最終確認
from sklearn.metrics import roc_auc_score
lgb_slim_auc = roc_auc_score(y, oof_lgb_slim)
print(f"\nLGBM Overall CV: {lgb_slim_auc:.5f}")

# 4. ベースラインとの比較表示
baseline_auc = 0.77854
diff = lgb_slim_auc - baseline_auc
print(f"ベースライン（初期）からの変化: {diff:+.5f}")

学習に使用するカラム数: 608
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[289]	training's auc: 0.85701	valid_1's auc: 0.778357
LGBM Fold 1 AUC: 0.77836
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[656]	training's auc: 0.916447	valid_1's auc: 0.779893
LGBM Fold 2 AUC: 0.77989
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[670]	training's auc: 0.917109	valid_1's auc: 0.77788
LGBM Fold 3 AUC: 0.77788
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[372]	training's auc: 0.873029	valid_1's auc: 0.780169
LGBM Fold 4 AUC: 0.78017
Training until validation scores don't improve for 100 rounds
Early stopping, best iteration is:
[467]	training's auc: 0.891965	valid_1's auc: 0.778099
LGBM Fold 5 AUC: 0.77810
LGBM Overall CV: 0.77875

[Slim版] LGBM Overall CV: 0.77875
ベースライン（初期）からの変化: +0.00021


【削除前】LGBM Overall CV: 0.77875  
【削除後】LGBM Overall CV: 0.77875  
変化なし

FE1結果まとめ
- 効いたFE：EXT_SOURCES_MEAN(LGBM1位)、GEOM_MEAN(CAT1位)、ORG_TYPE_TARGET_RATE(両方TOP5)
- 効かなかったFE：abnormal_MAX_frag(54カラム中ほぼ全滅)、round_1000k、border_frag
- AUCが微増に留まった原因：EXT_SOURCE合成特徴量が元のEXT_SOURCE_1〜3と食い合い、
  新しい情報を追加できていない。FE2ではEXT_SOURCEと独立した情報源のFEで上積みを狙う。  

参考【EXT_SOURCEの寄与度遷移(CAT)】  
EXT_SOURCE1:5.1853　→0.968648  
EXT_SOURCE2:16.1962 →1.37846  
EXT_SOURCE3:16.7332 →1.38394  
※クリーニング時はホールドアウト方式  

In [ ]:
#6-1　出力　parquet(zstd形式)

import os
# 1. zstd圧縮で保存
train_filename = 'train_FE1.zstd.parquet'
test_filename = 'test_FE1.zstd.parquet'

train.to_parquet(train_filename, compression='zstd', index=False)
test.to_parquet(test_filename, compression='zstd', index=False)

# 2. 保存したファイルのサイズを確認
# 変数名で参照することで書き換え漏れを防ぎます
train_size_zstd = os.path.getsize(train_filename) / 1024**2
test_size_zstd = os.path.getsize(test_filename) / 1024**2

print(f"Trainファイルサイズ (zstd): {train_size_zstd:.2f} MB")
print(f"Test ファイルサイズ (zstd): {test_size_zstd:.2f} MB")

Trainファイルサイズ (zstd): 223.04 MB
Test ファイルサイズ (zstd): 42.13 MB


クリーニング時にダウンキャストした場合は、FE時に型が衝突するため注意。  
改善策として全てfloat(64)に戻すのがシンプルだが、メモリがセッションクラッシュする可能性がある。
その場合はNumpyで処理して最後に追加する方法が有効。